compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -0,0 +1,615 @@
+//! However, when `inner_func` is called from `entry_point_two`, the address for `THREE` is `M32836`.
+//!
+//! This pass will duplicate `inner_func` so that different functions are called by the different entry points.
+//! The test module for this pass can be referenced to see how this function duplication looks in SSA.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use fxhash::{FxHashMap as HashMap, FxHashSet as HashSet};
+
+use crate::ssa::{
+    ir::{
+        function::{Function, FunctionId},
+        instruction::Instruction,
+        value::Value,
+    },
+    Ssa,
+};
+
+use super::inlining::called_functions_vec;
+
+impl Ssa {
+    pub(crate) fn brillig_entry_point_analysis(mut self) -> Ssa {
+        if self.main().runtime().is_brillig() {
+            return self;
+        }
+
+        let brillig_entry_points = get_brillig_entry_points(&self.functions, self.main_id);
+        let entry_points = brillig_entry_points.keys().copied().collect::<HashSet<_>>();
+
+        let inner_call_to_entry_point = build_inner_call_to_entry_points(&brillig_entry_points);
+
+        let functions_to_clone_map =
+            build_functions_to_clone(&self.functions, inner_call_to_entry_point, entry_points);
+
+        let calls_to_update = build_calls_to_update(&mut self, functions_to_clone_map);
+
+        let mut new_functions_map = HashMap::default();
+        for (entry_point, inner_calls) in brillig_entry_points {
+            let new_entry_point =
+                new_functions_map.get(&entry_point).copied().unwrap_or(entry_point);
+
+            let function =
+                self.functions.get_mut(&new_entry_point).expect("ICE: Function does not exist");
+            update_function_calls(function, entry_point, &mut new_functions_map, &calls_to_update);
+
+            for inner_call in inner_calls {
+                let new_inner_call =
+                    new_functions_map.get(&inner_call).copied().unwrap_or(inner_call);
+
+                let function =
+                    self.functions.get_mut(&new_inner_call).expect("ICE: Function does not exist");
+                update_function_calls(
+                    function,
+                    entry_point,
+                    &mut new_functions_map,
+                    &calls_to_update,
+                );
+            }
+        }
+
+        self
+    }
+}
+
+fn build_functions_to_clone(
+    functions: &BTreeMap<FunctionId, Function>,
+    inner_call_to_entry_point: HashMap<FunctionId, Vec<FunctionId>>,
+    entry_points: HashSet<FunctionId>,
+) -> HashMap<FunctionId, Vec<(Function, FunctionId)>> {
+    let mut functions_to_clone_map: HashMap<FunctionId, Vec<(Function, FunctionId)>> =
+        HashMap::default();
+
+    let mut add_function_to_clone = |entry_point: FunctionId, inner_call: FunctionId| {
+        let cloned_function = Function::clone_no_id(&functions[&inner_call]);
```

Comment:
I don't think that we need to actually clone the function here. This should be able to be done inside of `build_calls_to_update`, which would make this function much easier to test.

Reply:
We can then just make use of `clone_with_id` inside of the closure there.

Reply:
It would be nice to have some tests on this to sanity check this function once the above check is made.

Reply:
I have switched to using `clone_with_id`. This does require an extra clone of the function to fetch the function for cloning inside of the `add_fn` closure. I agree the code is easier to follow now though.

Reply:
> It would be nice to have some tests on this to sanity check this function once the above check is made.

Are you referring to `build_functions_to_clone` or `build_calls_to_update`? Do you not feel this logic is pretty much covered by the existing unit tests or are you thinking of a more explicit test on one of those methods?

Reply:
This was for `build_functions_to_clone` but the new implementation is so simple that :shrug: